File: Exercise_2.py
# Time complexity: O(1) for all operations
# Space complexity: O(1) if we are not considering the actual hashmap itself, otherwise O(n).
# Does it work on Leetcode: Yes
# Did you face any problems while doing this: No

# Here we use a linear chaining implementation to solve this problem.
# The linked list has a mazimum size of 100 given the constraints of this problem.
# We also use a dummy node for easier operations on the linked list.

class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.next = None
        
class MyHashMap:

    def __init__(self):
        self.pArr = [None for _ in range(10000)]

    def _hash(self, key):
        return key % 10000
    
    def _search(self, head, key):
        prev, curr = head, head.next
        while curr:
            if curr.key == key:
                break
            prev = curr
            curr = curr.next
        return prev, curr        

    def put(self, key: int, value: int) -> None:
        h = self._hash(key)
        if not self.pArr[h]:
            self.pArr[h] = Node(-1, -1)
        prev, curr = self._search(self.pArr[h], key)
        if not curr:
            prev.next = Node(key, value)
        else:
            curr.val = value

    def get(self, key: int) -> int:
        h = self._hash(key)
        if not self.pArr[h]:
            return -1
        prev, curr = self._search(self.pArr[h], key)
        if not curr:
            return -1
        return curr.val
        
    def remove(self, key: int) -> None:
        h = self._hash(key)
        if not self.pArr[h]:
            return
        prev, curr = self._search(self.pArr[h], key)
        if not curr:
            return
        prev.next = curr.next    

File: Exercise_2.cpp
// Design HashMap
// Time Complexity : O(1)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


// Approach
// Will use Linkedlist and linear chaining. Create a hash() & find() to keep track of prev.
// then in all put, find remove we will have 3 conditions. 1)If LL is present no not.
// 2) if prev node found and we are at the last node. 3) if prev node found and next node is also exist.

class Node {
public:
    int key, val;
    Node *next;

    Node(int key, int val) {
        this->key = key;
        this->val = val;
        this->next = nullptr;
    }
};

class MyHashMap {
public:
    vector<Node*> nodes;
    MyHashMap() {
        nodes.resize(10000);
    }
    
    int hash1(int key) {
        return key % 10000;
    }
    
    Node* find(Node *head, int key) {
        Node *prev = nullptr;
        Node *curr = head;
        while(curr != nullptr && curr->key != key) {
            prev = curr;
            curr = curr->next;
        }
        return prev;
    }
    
    void put(int key, int value) {
        int idx = hash1(key);
        if(nodes[idx] == nullptr) {
            nodes[idx] = new Node(-1, -1);
        }
        Node *prev = find(nodes[idx], key);
        if(prev->next == nullptr) {
            prev->next = new Node(key, value);
        }
        else {
            prev->next->val = value;
        }
    }
    
    int get(int key) {
        int idx = hash1(key);
        if(nodes[idx] == nullptr) return -1;
        Node *prev = find(nodes[idx], key);
        if(prev->next == nullptr) return -1;
        return prev->next->val;
    }
    
    void remove(int key) {
        int idx = hash1(key);
        if(nodes[idx] == nullptr) return;
        Node *prev = find(nodes[idx], key);
        if(prev->next == nullptr) return;
        Node *toDel = prev->next;
        prev->next = prev->next->next;
        delete toDel;
    }
};